elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/filter/FilterTranslator.java

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright 2019, Yahoo Inc.
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.core.filter;
+
+import static com.yahoo.elide.core.filter.Operator.FALSE;
+import static com.yahoo.elide.core.filter.Operator.GE;
+import static com.yahoo.elide.core.filter.Operator.GT;
+import static com.yahoo.elide.core.filter.Operator.IN;
+import static com.yahoo.elide.core.filter.Operator.INFIX;
+import static com.yahoo.elide.core.filter.Operator.INFIX_CASE_INSENSITIVE;
+import static com.yahoo.elide.core.filter.Operator.IN_INSENSITIVE;
+import static com.yahoo.elide.core.filter.Operator.ISNULL;
+import static com.yahoo.elide.core.filter.Operator.LE;
+import static com.yahoo.elide.core.filter.Operator.LT;
+import static com.yahoo.elide.core.filter.Operator.NOT;
+import static com.yahoo.elide.core.filter.Operator.NOTNULL;
+import static com.yahoo.elide.core.filter.Operator.NOT_INSENSITIVE;
+import static com.yahoo.elide.core.filter.Operator.POSTFIX;
+import static com.yahoo.elide.core.filter.Operator.POSTFIX_CASE_INSENSITIVE;
+import static com.yahoo.elide.core.filter.Operator.PREFIX;
+import static com.yahoo.elide.core.filter.Operator.PREFIX_CASE_INSENSITIVE;
+import static com.yahoo.elide.core.filter.Operator.TRUE;
+
+import com.yahoo.elide.core.Path;
+import com.yahoo.elide.core.exceptions.InvalidPredicateException;
+import com.yahoo.elide.core.filter.FilterPredicate.FilterParameter;
+import com.yahoo.elide.core.filter.expression.AndFilterExpression;
+import com.yahoo.elide.core.filter.expression.FilterExpression;
+import com.yahoo.elide.core.filter.expression.FilterExpressionVisitor;
+import com.yahoo.elide.core.filter.expression.NotFilterExpression;
+import com.yahoo.elide.core.filter.expression.OrFilterExpression;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.lang3.tuple.Triple;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Translates a filter predicate into a JPQL fragment.
+ */
+public class FilterTranslator implements FilterOperation<String> {
+    private static final String COMMA = ", ";
+
+    private static Map<Operator, JPQLPredicateGenerator> operatorGenerators;
+    private static Map<Triple<Operator, Class<?>, String>, JPQLPredicateGenerator> predicateOverrides;
+
+    /**
+     * Converts a JPQL column alias and list of arguments into a JPQL filter predicate fragment.
+     */
+    @FunctionalInterface
+    public interface JPQLPredicateGenerator {
+        String generate(String columnAlias, List<FilterParameter> parameters);
+    }
+
+    static {
+        predicateOverrides = new HashMap<>();
```

Comment:
I'm not sure if I understand how `predicateOverrides` will be used, because looks like all useful generators are in `operatorGenerators`.

Reply:
PredicateOverrides is useful if you have a table with a different case sensitivity strategy from the rest of the database.  You can generate different JPQL just for that table.